tfjs-backend-webgl/src/backend_webgl.ts

```diff
@@ -380,6 +380,53 @@ export class MathBackendWebGL extends KernelBackend {
     return dTypeVals;
   }
 
+  /**
+   * Read tensor to a new texture that is densely packed for ease of use.
+   * @param dataId The source tensor.
+   * @param options
+   *     customTexShape: Optional. If set, will use the user defined texture
+   *     shape to create the texture.
+   */
+  readToGPU(dataId: DataId, options: DataToGPUWebGLOption = {}): GPUResource {
+    const texData = this.texData.get(dataId);
+    const {values, shape, slice, dtype, isPacked, texture} = texData;
+
+    if (dtype === 'complex64') {
+      throw new Error('Does not support reading texture for complex64 dtype.');
+    }
+
+    // The presence of `slice` indicates this tensor is a shallow slice of a
+    // different tensor, and is using that original tensor's texture. Run
+    // `clone` in order to copy that texture and read from it.
+    if (slice != null) {
+      let program;
+      if (isPacked) {
+        program = new UnaryOpPackedProgram(shape, unary_op.CLONE);
+      } else {
+        program = new UnaryOpProgram(shape, unary_op.CLONE);
+      }
+      const res =
+          this.runWebGLProgram(program, [{dataId, shape, dtype}], dtype);
+      const gpuResouorce = this.readToGPU(res, options);
+      this.disposeIntermediateTensorInfo(res);
+      return gpuResouorce;
+    }
+
+    if (values != null && texture == null) {
```

Comment:
If `values` is not null, can we just upload the data to a texture and return the texture?